### DISCUSSION TITLE
Adding ability to have margin between tiles in inventory grid and texture background at the tile

### DISCUSSION THREAD

**Discussion by filipinyo** [peter-kish/gloot#41] [Ideas]
Currently there are only lines between the tiles, so as stated in title, it would be useful to have the ability to set the spacing between the tiles in the inventory grid. It would also be handy if one could add a background texture for the tiles, preferably texture for different states:

-  Normal state 
-  Highlighted (when the user hover's over it)
-  Selected (when the user selected it)
-  Disabled (can't be selected etc)

adding the ability to separate tiles with textures instead of lines would increase the customizability a lot. Since a lot of Godot users make pixel art games (myself included), non pixel lines aren't always the preferred  style choice. I know that currently this would be a challenge to implement with the items that have size larger that 1x1 tile, maybe we could create a separate inventory type for that, which only allows items 1x1 and name it something else? Maybe inventory_retro_grid etc.

Looking forward for your opinion.

**Comment by peter-kish** [comment]
So far I tried to avoid making the UI control classes "too customizable", simply because a UI is something very specific for each game and I was afraid that increasing customizability to cover as many of those cases as possible could overcomplicate those classes (apart from being a lot of work).

Because if that I was mostly focusing on making some very basic UI control classes that are easy to understand and that could easily be extended or build on top of by the user (I was even thinking about making the grid lines optional). Borders, background images, labels etc. I meant to be added in combination with other control nodes (`TextureRect`, `ColorRect`, `Label` etc.).

But seeing that a lot of people are using the plugin exactly because of the UI control nodes (especially `CtrlInventoryGrid`) I think adding something like a `CtrInventoryGridEx` (extended) node could be a compromise. It could support some additional customization, while `CtrlInventoryGrid` would stay "barebones" enough for anyone who wants to build something unique on top of it.

**Comment by peter-kish** [comment]
Spacing between items, however, is somewhat different because it impacts the way field coordinates are calculated. I think it's something that's fairly commonly used but would be difficult to implement on top of `CtrlInventoryGrid` and now I'm seriously considering to add it as an option.
As you already noted, displaying items that are larger than 1x1 tile would be somewhat of a challenge (the width of a 3x3 item is no longer `3 * field_dimensions.x`, but `3 * field_dimensions.x + 2 * spacing`) and the artwork would need be sized accordingly.

**Comment by peter-kish** [comment]
> maybe we could create a separate inventory type for that, which only allows items 1x1 and name it something else? Maybe inventory_retro_grid etc.

Not sure if this would be necessary as there's nothing preventing you to only define items that are 1x1 in size.

**Comment by peter-kish** [comment]
Item spacing has been added in [v2.0.3](https://github.com/peter-kish/gloot/releases/tag/v2.0.3) and `CtrInventoryGridEx` has been added in https://github.com/peter-kish/gloot/pull/53. It has not been tested thoroughly so please report any potential issues.

**Reply by filipinyo** [reply]
Great work! It works good!

![image](https://user-images.githubusercontent.com/6736045/194758199-b56fe5fa-18d2-47d0-86b2-1754bac6021e.png)

**Reply by filipinyo** [reply]
Occasionaly this error comes up:

![image](https://user-images.githubusercontent.com/6736045/194758570-812db7c2-2545-4873-b35a-de974e266ffc.png)

and also this one that items overlap:

![image](https://user-images.githubusercontent.com/6736045/194758734-6d5bec43-f3ec-4efd-af54-ab4714568c00.png)
 This last one could be a problem if you remove the items in the scene or something like that?

**Reply by filipinyo** [reply]
Little bi more of the debugger errors:

![image](https://user-images.githubusercontent.com/6736045/194760415-73f47dae-a2bf-4563-9d6d-4d6663523228.png)

seems like it goes in the infinite loop occassionaly.

**Reply by peter-kish** [reply]
The stack overflow issue should be fixed now (a0f6ce9def81596fe5cd141b5e40cc7a2d275dee). Still haven't figured out the overlapping items issue though.

**Reply by filipinyo** [reply]
@peter-kish thank you! I'll test it out tomorrow and I'll look in more details what happens with the overlapping items over the weekend. One thing that I've noticed though... Sometimes (really not always) if I swap items, example I put the item from position 6 to position 2 the item on position 4 gets placed on other position. I'll try to replicate how I trigger this on the weekend.

**Reply by peter-kish** [reply]
I played around with it a little, but couldn't replicate the issue. If you manage to figure out how to do it, please let me know.

So far, I can only assume that it has something to do with the `_sort_if_needed()` method inside `InventoryGrid`. You can also try commenting out the two calls to `_sort_if_needed()` inside `inventory_grid.gd` and check if it the bug persists.

I might remove this "auto sorting" feature at some point as it seems to introduce more problems than it solves.

**Reply by filipinyo** [reply]
@peter-kish The assertion is fixed. But sorting is still messed up even if I comment out _sort_if_needed(). I have 2 units that have their own inventories and items in them (seperate instances), I select first unit and rearange it's items, then I select another unit with inventory and it has the items arranged in the same way as first unit. Will post more if I find the reasons etc.
